<launch>

    <node name="map_to_odom" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 1 map odom"/>
    <node name="base_to_laser" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 -0.044 0.999 base_footprint laser_frame"/>

    <node name="nodelet_manager" pkg="nodelet" type="nodelet" args="manager" output="screen">
        <param name="num_worker_threads" value="20"/>
    </node>

    <node name="scans_merger" pkg="nodelet" type="nodelet" args="load obstacle_detector/ScansMerger nodelet_manager">
        <param name="active" value="true"/>
        <param name="publish_scan" value="true"/>
        <param name="publish_pcl" value="false"/>

        <param name="ranges_num" value="1000"/>

        <param name="min_scanner_range" value="0.05"/>
        <param name="max_scanner_range" value="10.0"/>

        <param name="min_x_range" value="-10.0"/>
        <param name="max_x_range" value="10.0"/>
        <param name="min_y_range" value="-10.0"/>
        <param name="max_y_range" value="10.0"/>

        <param name="fixed_frame_id" value="map"/>
        <param name="target_frame_id" value="odom"/>

        <remap from="scan" to="scan_map"/>
    </node>

    <node name="obstacle_extractor" pkg="nodelet" type="nodelet" args="load obstacle_detector/ObstacleExtractor nodelet_manager">
        <param name="active" value="true"/>
        <param name="use_scan" value="true"/>
        <param name="use_pcl" value="false"/>

        <param name="use_split_and_merge" value="true"/>
        <param name="circles_from_visibles" value="true"/>
        <param name="discard_converted_segments" value="false"/>
        <param name="transform_coordinates" value="true"/>

        <param name="min_group_points" value="2"/>

        <param name="max_group_distance" value="0.04"/>
        <param name="distance_proportion" value="0.00628"/>
        <param name="max_split_distance" value="0.02"/>
        <param name="max_merge_separation" value="0.1"/>
        <param name="max_merge_spread" value="0.02"/>
        <param name="max_circle_radius" value="0.2"/>
        <param name="radius_enlargement" value="0.05"/>

        <param name="frame_id" value="laser_frame"/>
    </node>

    <!-- <node name="obstacle_tracker" pkg="nodelet" type="nodelet" args="load obstacle_detector/ObstacleTracker nodelet_manager">
        <param name="active" value="true"/>

        <param name="loop_rate" value="12.0"/>
        <param name="tracking_duration" value="0.5"/>
        <param name="min_correspondence_cost" value="0.6"/>
        <param name="std_correspondence_dev" value="0.15"/>
        <param name="process_variance" value="0.1"/>
        <param name="process_rate_variance" value="0.1"/>
        <param name="measurement_variance" value="1.0"/>

        <param name="frame_id" value="laser_frame"/>

        <remap from="tracked_obstacles" to="obstacles"/>
    </node> -->

    <node name="lidar_localization" pkg="nodelet" type="nodelet" args="load lidar_localization/LidarLocalization nodelet_manager" output="log">
        <param name="active" value="true"/>

        <param name="cov_x" value="0.9"/>
        <param name="cov_y" value="0.9"/>
        <param name="cov_yaw" value="0.9"/>

        <param name="beacon_1_x" value="1"/>
        <param name="beacon_1_y" value="-0.08"/>
  
        <param name="beacon_2_x" value="0.08"/>
        <param name="beacon_2_y" value="3.18"/>
  
        <param name="beacon_3_x" value="1.92"/>
        <param name="beacon_3_y" value="3.18"/>
        <param name="theta" value="4"/>

        <param name="obstacle_topic" value="raw_obstacles"/>
        <param name="beacon_parent_frame_id" value="map"/>
        <param name="beacon_frame_id_prefix" value="beacon"/>
        <param name="robot_parent_frame_id" value="map"/>
        <param name="robot_frame_id" value="base_footprint"/>
    </node>

    <node name="rx_to_odom" pkg="lidar_localization" type="rx_to_odom">
        <param name="active" value="true"/>
        <param name="publish_topic" value="true"/>
        <param name="init_x" value="1.65"/>
        <param name="init_y" value="2.58"/>
        <param name="init_yaw" value="-2.66"/>
        <param name="cov_vx" value="0.05"/>
        <param name="cov_vyaw" value="0.05"/>
        <param name="delay" value="1"/>
    </node>

</launch>
